File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/aas/impl/metamodel/hashmap/aas/SubModel.java
package org.eclipse.basyx.aas.impl.metamodel.hashmap.aas;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import org.eclipse.basyx.aas.api.metamodel.aas.ISubModel;
import org.eclipse.basyx.aas.api.metamodel.aas.identifier.IIdentifier;
import org.eclipse.basyx.aas.api.metamodel.aas.qualifier.IAdministrativeInformation;
import org.eclipse.basyx.aas.api.metamodel.aas.reference.IReference;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.IDataElement;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.ISubmodelElement;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.operation.IOperation;
import org.eclipse.basyx.aas.api.metamodel.aas.submodelelement.property.IProperty;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasDataSpecificationFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasKindFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.HasSemanticsFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.IdentifiableFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.ReferableFacade;
import org.eclipse.basyx.aas.impl.metamodel.facades.SubmodelFacade;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.IVABElementContainer;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.VABModelMap;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.HasDataSpecification;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.HasSemantics;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.Identifiable;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.haskind.HasKind;
import org.eclipse.basyx.aas.impl.metamodel.hashmap.aas.qualifier.qualifiable.Qualifiable;

/**
 * Submodel class
 * 
 * @author kuhn, schnicke
 *
 *
 */
public class SubModel extends VABModelMap<Object> implements IVABElementContainer, ISubModel {

	/**
	 * Version of serialized instances
	 */
	private static final long serialVersionUID = 1L;

	public static final String SUBMODELELEMENT = "submodelElement";
	public static final String PROPERTIES = "dataElements";
	public static final String OPERATIONS = "operations";

	/**
	 * Submodel properties
	 */
	protected Map<String, IDataElement> dataElements = new HashMap<>();

	/**
	 * Submodel operations
	 */
	protected Map<String, IOperation> operations = new HashMap<>();

	/**
	 * Submodel elements in general. Does also contain operations and properties
	 */
	protected Map<String, ISubmodelElement> elements = new HashMap<>();

	/**
	 * Constructor
	 */
	public SubModel() {
		// Add qualifiers
		putAll(new HasSemantics());
		putAll(new Identifiable());
		putAll(new Qualifiable());
		putAll(new HasDataSpecification());
		putAll(new HasKind());

		// Attributes
		put(SUBMODELELEMENT, elements);

		// Helper attributes
		put(PROPERTIES, dataElements);
		put(OPERATIONS, operations);
	}

	/**
	 * Constructor
	 */
	public SubModel(HasSemantics semantics, Identifiable identifiable, Qualifiable qualifiable, HasDataSpecification specification, HasKind hasKind) {
		// Add qualifiers
		putAll(semantics);
		putAll(identifiable);
		putAll(qualifiable);
		putAll(specification);
		putAll(hasKind);

		// Attributes
		put(SUBMODELELEMENT, elements);

		// Helper attributes
		put(PROPERTIES, dataElements);
		put(OPERATIONS, operations);
	}

	@Override
	public IReference getSemanticId() {
		return new HasSemanticsFacade(this).getSemanticId();
	}

	public void setSemanticID(IReference ref) {
		new HasSemanticsFacade(this).setSemanticID(ref);
	}

	@Override
	public IAdministrativeInformation getAdministration() {
		return new IdentifiableFacade(this).getAdministration();
	}

	@Override
	public IIdentifier getIdentification() {
		return new IdentifiableFacade(this).getIdentification();
	}

	public void setAdministration(String version, String revision) {
		new IdentifiableFacade(this).setAdministration(version, revision);
	}

	public void setIdentification(String idType, String id) {
		new IdentifiableFacade(this).setIdentification(idType, id);
	}

	@Override
	public Set<IReference> getDataSpecificationReferences() {
		return new HasDataSpecificationFacade(this).getDataSpecificationReferences();
	}

	public void setDataSpecificationReferences(Set<IReference> ref) {
		new HasDataSpecificationFacade(this).setDataSpecificationReferences(ref);
	}

	@Override
	public String getHasKindReference() {
		return new HasKindFacade(this).getHasKindReference();
	}

	public void setHasKindReference(String kind) {
		new HasKindFacade(this).setHasKindReference(kind);
	}

	@Override
	public String getId() {
		return new SubmodelFacade(this).getId();
	}

	@Override
	public void setId(String id) {
		new SubmodelFacade(this).setId(id);
	}

	@Override
	public void setProperties(Map<String, IProperty> properties) {
		new SubmodelFacade(this).setProperties(properties);
	}

	@Override
	public void setOperations(Map<String, IOperation> operations) {
		new SubmodelFacade(this).setOperations(operations);
	}

	@Override
	public String getIdshort() {
		return new ReferableFacade(this).getIdshort();
	}

	@Override
	public String getCategory() {
		return new ReferableFacade(this).getCategory();
	}

	@Override
	public String getDescription() {
		return new ReferableFacade(this).getDescription();
	}

	@Override
	public IReference getParent() {
		return new ReferableFacade(this).getParent();
	}

	public void setIdshort(String idShort) {
		new ReferableFacade(this).setIdshort(idShort);
	}

	public void setCategory(String category) {
		new ReferableFacade(this).setCategory(category);
	}

	public void setDescription(String description) {
		new ReferableFacade(this).setDescription(description);
	}

	public void setParent(IReference obj) {
		new ReferableFacade(this).setParent(obj);
	}

	@Override
	public Map<String, IDataElement> getDataElements() {
		return new SubmodelFacade(this).getDataElements();
	}

	@Override
	public Map<String, IOperation> getOperations() {
		return new SubmodelFacade(this).getOperations();
	}

	@Override
	public void addSubModelElement(ISubmodelElement element) {
		new SubmodelFacade(this).addSubModelElement(element);
	}
}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/simple/SubModel.cpp
#include <BaSyx/submodel/simple/SubModel.h>

using namespace basyx::submodel;
using namespace basyx::submodel::api;
using namespace basyx::submodel::simple;

SubModel::SubModel(const std::string & idShort, const Identifier & identifier)
	: kind(ModelingKind::Instance)
	, identifiable(idShort, identifier)
{
};

IElementContainer<ISubmodelElement> & SubModel::submodelElements()
{
	return this->elementContainer;
}

const IElementContainer<ISubmodelElement> & SubModel::submodelElements() const
{
	return this->elementContainer;
}

ModelingKind SubModel::getKind() const
{
	return this->kind;
}

const std::string & SubModel::getIdShort() const
{
	return this->identifiable.getIdShort();
}

const std::string * const SubModel::getCategory() const
{
	return this->identifiable.getCategory();
};

void SubModel::setCategory(const std::string & category)
{
	this->identifiable.setCategory(category);
};

LangStringSet & SubModel::getDescription()
{
	return this->identifiable.getDescription();
}

const LangStringSet & SubModel::getDescription() const
{
	return this->identifiable.getDescription();
}

IReferable * SubModel::getParent() const
{
	return this->identifiable.getParent();
}

const AdministrativeInformation & SubModel::getAdministrativeInformation() const
{
	return this->identifiable.getAdministrativeInformation();
}

Identifier SubModel::getIdentification() const
{
	return this->identifiable.getIdentification();
}

AdministrativeInformation & SubModel::getAdministrativeInformation()
{
	return this->identifiable.getAdministrativeInformation();
}

bool SubModel::hasAdministrativeInformation() const
{
	return this->identifiable.hasAdministrativeInformation();
};

void SubModel::addDataSpecification(const Reference & reference)
{
	this->dataSpecification.addDataSpecification(reference);
}

const std::vector<Reference> SubModel::getDataSpecificationReference() const
{
	return this->dataSpecification.getDataSpecificationReference();
}

const IReference & SubModel::getSemanticId() const
{
	return this->semanticId;
}

void SubModel::setSemanticId(const IReference & semanticId)
{
	this->semanticId = semanticId;
}

void SubModel::addFormula(const api::IFormula & formula)
{
	this->qualifiable.addFormula(formula);
};

void SubModel::addQualifier(const api::IQualifier & qualifier)
{
	this->qualifiable.addQualifier(qualifier);
};

std::vector<simple::Formula> SubModel::getFormulas() const
{
	return this->qualifiable.getFormulas();
};

std::vector<simple::Qualifier> SubModel::getQualifiers() const
{
	return this->qualifiable.getQualifiers();
};

